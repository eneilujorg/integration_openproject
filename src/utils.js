let mytimer = 0
export function delay(callback, ms) {
	return function() {
		const context = this
		const args = arguments
		clearTimeout(mytimer)
		mytimer = setTimeout(function() {
			callback.apply(context, args)
		}, ms || 0)
	}
}

export const STATE = {
	OK: 'ok',
	ERROR: 'error',
	LOADING: 'loading',
	NO_TOKEN: 'no-token',
	CONNECTION_ERROR: 'connection-error',
	FAILED_FETCHING_WORKPACKAGES: 'failed-fetching-workpackages',
}
