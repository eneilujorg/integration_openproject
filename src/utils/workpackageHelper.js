import { generateUrl } from '@nextcloud/router'
import axios from '@nextcloud/axios'

export const workpackageHelper = {
	async getColorAttributes(path, id) {
		const url = generateUrl(path + id)
		let response
		try {
			response = await axios.get(url)
		} catch (e) {
			response = e.response
		}
		return (response.status === 200 && response.data?.color)
			? response.data.color
			: ''
	},
	replaceHrefToGetId(href) {
		// this is a helper method replaces the string like this "/api/v3/types/3" to get id
		return href
			? href.replace(/.*\//, '')
			: null
	},
	async getAdditionalMetaData(workPackage) {
		if (typeof workPackage._links.status.href !== 'string'
			|| workPackage._links.status.href === ''
			|| typeof workPackage._links.type.href !== 'string'
			|| workPackage._links.type.href === ''
			|| typeof workPackage.id !== 'number'
			|| typeof workPackage.subject !== 'string'
			|| workPackage.subject === ''
			|| typeof workPackage._links.project.title !== 'string'
			|| workPackage._links.project.title === ''
			|| typeof workPackage._links.status.title !== 'string'
			|| workPackage._links.status.title === ''
			|| typeof workPackage._links.type.title !== 'string'
			|| workPackage._links.type.title === ''
		) {
			throw new Error('missing data in workpackage object')
		}
		const statusId = this.replaceHrefToGetId(workPackage._links.status.href)
		const typeId = this.replaceHrefToGetId(workPackage._links.type.href)
		const userId = this.replaceHrefToGetId(workPackage._links.assignee.href)
		const projectId = this.replaceHrefToGetId(workPackage._links.project.href)
		const userName = workPackage._links.assignee.title
		const avatarUrl = generateUrl('/apps/integration_openproject/avatar?')
			+ encodeURIComponent('userId')
			+ '=' + userId
			+ '&' + encodeURIComponent('userName')
			+ '=' + userName
		const statusColor = await this.getColorAttributes('/apps/integration_openproject/statuses/', statusId)
		const typeColor = await this.getColorAttributes('/apps/integration_openproject/types/', typeId)
		return {
			id: workPackage.id,
			subject: workPackage.subject,
			project: workPackage._links.project.title,
			projectId,
			statusTitle: workPackage._links.status.title,
			typeTitle: workPackage._links.type.title,
			assignee: userName,
			statusCol: statusColor,
			typeCol: typeColor,
			picture: avatarUrl,
		}
	},
}
