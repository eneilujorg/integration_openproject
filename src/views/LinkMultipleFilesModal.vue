<template>
	<div class="multiple-link-modal-container">
		<NcModal
			v-if="show"
			@close="closeRequestModal">
			<div class="multiple-link-modal-content">
				<LoadingIcon v-if="isLoading" class="loading-spinner" :size="60" />
				<div v-else class="multiple-link-modal-inside-content">
					<h2>
						{{ t('integration-openproject', 'Link to work package') }}
					</h2>
					<SearchInput v-if="!!isAdminConfigOk && !!isStateOk"
						:linked-work-packages="alreadyLinkedWorkPackage"
						:file-info="fileInfos"
						:search-origin="searchOrigin"
						@saved="onSaved" />
					<EmptyContent
						id="openproject-empty-content"
						:state="state"
						:is-multiple-workpackage-linking="true"
						:is-admin-config-ok="isAdminConfigOk" />
				</div>
			</div>
		</NcModal>
	</div>
</template>

<script>
import NcModal from '@nextcloud/vue/dist/Components/NcModal.js'
import SearchInput from '../components/tab/SearchInput.vue'
import EmptyContent from '../components/tab/EmptyContent.vue'
import { generateUrl } from '@nextcloud/router'
import axios from '@nextcloud/axios'
import { loadState } from '@nextcloud/initial-state'
import LoadingIcon from 'vue-material-design-icons/Loading.vue'
import { WORKPACKAGES_SEARCH_ORIGIN, STATE } from '../utils.js'
import { workpackageHelper } from '../utils/workpackageHelper.js'

export default {
	name: 'LinkMultipleFilesModal',
	components: {
		EmptyContent,
		SearchInput,
		NcModal,
		LoadingIcon,
	},
	data() {
		return {
			show: false,
			state: STATE.LOADING,
			fileInfos: [],
			alreadyLinkedWorkPackage: [],
			isAdminConfigOk: loadState('integration_openproject', 'admin-config-status'),
			searchOrigin: WORKPACKAGES_SEARCH_ORIGIN.LINK_MULTIPLE_FILES_MODAL,
		}
	},

	computed: {
		isStateOk() {
			return this.state === STATE.OK
		},
		isLoading() {
			return this.state === STATE.LOADING
		},
	},
	methods: {
		showModal() {
			this.show = true
		},
		closeDropDown() {
			this.show = false
		},
		async setFileInfos(fileInfos) {
			this.fileInfos = fileInfos
			if (this.isAdminConfigOk) {
				await this.fetchWorkpackagesForSingleFileSelected(this.fileInfos[0].id)
			} else {
				this.state = STATE.ERROR
			}
		},
		closeRequestModal() {
			this.fileInfos = []
			this.alreadyLinkedWorkPackage = []
			this.show = false
		},
		async fetchWorkpackagesForSingleFileSelected(fileId) {
			this.state = STATE.LOADING
			const req = {}
			const url = generateUrl('/apps/integration_openproject/work-packages?fileId=' + fileId)
			try {
				const response = await axios.get(url, req)
				if (!Array.isArray(response.data)) {
					this.state = STATE.FAILED_FETCHING_WORKPACKAGES
				} else {
					if (this.fileInfos.length === 1 && response.data.length > 0) {
						for (let workPackage of response.data) {
							workPackage.fileId = fileId
							workPackage = await workpackageHelper.getAdditionalMetaData(workPackage, true)
							this.alreadyLinkedWorkPackage.push(workPackage)
						}
					}
					this.state = STATE.OK
				}
			} catch (error) {
				if (error.response && error.response.status === 401) {
					this.state = STATE.NO_TOKEN
				} else if (error.response && (error.response.status === 404 || error.response.status === 503)) {
					this.state = STATE.CONNECTION_ERROR
				} else if (error.response && error.response.status === 500) {
					this.state = STATE.ERROR
				} else {
					this.state = STATE.FAILED_FETCHING_WORKPACKAGES
				}
			}
		},
		onSaved() {
			this.closeRequestModal()
		},
	},
}
</script>

<style scoped lang="scss">
h2 {
	padding-left: 15px;
}

.multiple-link-modal-content {
	padding: 20px 30px;
	height: 450px;
	display: flex;
	justify-content: center;
	align-items: center;
}

.multiple-link-modal-inside-content {
	width: 100%;
	height: 100%;
	display: flex;
	justify-content: center;
	flex-direction: column;
}
</style>
