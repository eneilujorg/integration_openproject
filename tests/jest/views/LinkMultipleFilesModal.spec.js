/* jshint esversion: 8 */

import { mount, createLocalVue, shallowMount } from '@vue/test-utils'
import LinkMultipleFilesModal from '../../../src/views/LinkMultipleFilesModal.vue'
import * as initialState from '@nextcloud/initial-state'
import { WORKPACKAGES_SEARCH_ORIGIN, STATE } from '../../../src/utils.js'
import { workpackageHelper } from '../../../src/utils/workpackageHelper.js'
import axios from '@nextcloud/axios'
import { getCurrentUser } from '@nextcloud/auth'

jest.mock('@nextcloud/auth')
jest.mock('@nextcloud/axios')
jest.mock('@nextcloud/dialogs')
const localVue = createLocalVue()

const singleFileInfo = [{
	id: 123,
	name: 'logo.ong',
}]

const multipleFileInfo = [{
	id: 123,
	name: 'logo.ong',
},

{
	id: 456,
	name: 'pogo.ong',
},
{
	id: 789,
	name: 'togo.ong',
}]

describe('LinkMultipleFilesModal.vue', () => {
	const searchInputStubSelector = 'searchinput-stub'
	const ncModalStubSelector = 'ncmodal-stub'
	const loadingIndicatorSelector = '.loading-spinner'
	const emptyContentSelector = '#openproject-empty-content'
	const emptyContentTitleMessageSelector = '.empty-content--message--title'
	beforeEach(() => {
		jest.useFakeTimers()
		// eslint-disable-next-line no-import-assign,import/namespace
		initialState.loadState = jest.fn(() => true)
	})

	describe('modal', () => {
		it('should open when "show" is set to true', async () => {
			const wrapper = shallowMount(LinkMultipleFilesModal, { localVue })
			await wrapper.setData({
				show: false,
			})
			await localVue.nextTick()
			await wrapper.vm.showModal()
			expect(wrapper.find(ncModalStubSelector).exists()).toBeTruthy()
		})

		it('should close when "show" is set to false', async () => {
			const wrapper = shallowMount(LinkMultipleFilesModal, { localVue })
			await localVue.nextTick()
			await wrapper.vm.closeRequestModal()
			expect(wrapper.find(ncModalStubSelector).exists()).toBeFalsy()
		})
	})

	describe('search input existence in modal', () => {
		let wrapper
		beforeEach(() => {
			wrapper = mountWrapper()
		})
		it('should not exist if admin config is not ok', async () => {
			await wrapper.setData({
				state: STATE.OK,
				isAdminConfigOk: false,
			})
			expect(wrapper.find(searchInputStubSelector).exists()).toBeFalsy()
		})
		it.each([
			{ state: STATE.NO_TOKEN },
			{ state: STATE.CONNECTION_ERROR },
			{ state: STATE.ERROR },
		])('should not exist if the wrapper is not in "ok" state', async (cases) => {
			await wrapper.setData({
				state: cases.STATE,
			})
			expect(wrapper.find(searchInputStubSelector).exists()).toBeFalsy()
		})

		it('should exist if the admin config is ok and the wrapper is in "ok" state', async () => {
			await wrapper.setData({
				isAdminConfigOk: true,
				state: STATE.OK,
			})
			expect(wrapper.find(searchInputStubSelector).exists()).toBeTruthy()
		})
	})

	describe('loading icon', () => {
		let wrapper
		beforeEach(() => {
			wrapper = mountWrapper()
		})
		it('should show the loading icon during "loading" state', async () => {
			await wrapper.setData({ state: STATE.LOADING })
			await localVue.nextTick()
			expect(wrapper.find(loadingIndicatorSelector).exists()).toBeTruthy()
		})
		it('should not show the empty content message during "loading" state', async () => {
			await wrapper.setData({ state: STATE.LOADING })
			await localVue.nextTick()
			expect(wrapper.find(emptyContentSelector).exists()).toBeFalsy()
		})
		it.each([STATE.OK, STATE.ERROR])('should make the loading icon disappear on state change', async (state) => {
			await wrapper.setData({ state: STATE.LOADING })
			await localVue.nextTick()
			expect(wrapper.find(loadingIndicatorSelector).exists()).toBeTruthy()
			await wrapper.setData({ state })
			await localVue.nextTick()
			expect(wrapper.find(loadingIndicatorSelector).exists()).toBeFalsy()
		})
	})

	describe('empty content', () => {
		let wrapper
		beforeEach(() => {
			wrapper = mountWrapper()
		})
		it.each([STATE.NO_TOKEN, STATE.ERROR, STATE.OK])('shows the empty message when state is other than loading', async (state) => {
			await wrapper.setData({ state })
			await localVue.nextTick()
			expect(wrapper.find(emptyContentSelector).exists()).toBeTruthy()
		})

		it('shows message "Add a new link to all selected files" when admin config is okay', async () => {
			wrapper = mount(LinkMultipleFilesModal, {
				localVue,
				attachTo: document.body,
				mocks: {
					t: (app, msg) => msg,
					generateUrl() {
						return '/'
					},
				},
				stubs: {
					SearchInput: true,
					NcModal: true,
				}
			})
			await wrapper.setData({
				show: true,
				state: STATE.OK,
				isAdminConfigOk: true
			})
			expect(wrapper.find(emptyContentSelector).exists()).toBeTruthy()
			const titleContent = wrapper.find(emptyContentTitleMessageSelector)
			expect(titleContent.text()).toBe('Add a new link to all selected files')
		})
	})

	describe('fetch workpackages', () => {
		let wrapper
		let axiosGetSpy = jest.fn()
		beforeEach(() => {
			wrapper = mountWrapper()
			axiosGetSpy.mockRestore()
			workpackageHelper.clearCache()
		})

		describe('single file selected', () => {
			it.each([
				{ HTTPStatus: 400, AppState: STATE.FAILED_FETCHING_WORKPACKAGES },
				{ HTTPStatus: 401, AppState: STATE.NO_TOKEN },
				{ HTTPStatus: 402, AppState: STATE.FAILED_FETCHING_WORKPACKAGES },
				{ HTTPStatus: 404, AppState: STATE.CONNECTION_ERROR },
				{ HTTPStatus: 500, AppState: STATE.ERROR },
			])('sets states according to HTTP error codes', async (cases) => {
				const err = new Error()
				err.response = { status: cases.HTTPStatus }
				axios.get.mockRejectedValueOnce(err)
				await wrapper.vm.setFileInfos(singleFileInfo)
				expect(wrapper.vm.state).toBe(cases.AppState)
			})

			it.each([
				null,
				'string',
				undefined,
				[{ // missing id
					subject: 'subject',
					_links: {
						status: {
							href: '/api/v3/statuses/12',
							title: 'open',
						},
						type: {
							href: '/api/v3/types/6',
							title: 'Task',
						},
						assignee: {
							href: '/api/v3/users/1',
							title: 'Bal Bahadur Pun',
						},
						project: { title: 'a big project' },
					},
				}],
				[{ // empty subject
					id: 123,
					subject: '',
					_links: {
						status: {
							href: '/api/v3/statuses/12',
							title: 'open',
						},
						type: {
							href: '/api/v3/types/6',
							title: 'Task',
						},
						assignee: {
							href: '/api/v3/users/1',
							title: 'Bal Bahadur Pun',
						},
						project: { title: 'a big project' },
					},
				}],
				[{ // missing subject
					id: 123,
					_links: {
						status: {
							href: '/api/v3/statuses/12',
							title: 'open',
						},
						type: {
							href: '/api/v3/types/6',
							title: 'Task',
						},
						assignee: {
							href: '/api/v3/users/1',
							title: 'Bal Bahadur Pun',
						},
						project: { title: 'a big project' },
					},
				}],
				[{ // missing _links.status.title
					id: 123,
					subject: 'my task',
					_links: {
						status: {
							href: '/api/v3/statuses/12',
						},
						type: {
							href: '/api/v3/types/6',
							title: 'Task',
						},
						assignee: {
							href: '/api/v3/users/1',
							title: 'Bal Bahadur Pun',
						},
						project: { title: 'a big project' },
					},
				}],
				[{ // missing project.title
					id: 123,
					subject: 'my task',
					_links: {
						status: {
							href: '/api/v3/statuses/12',
							title: 'open',
						},
						type: {
							href: '/api/v3/types/6',
							title: 'Task',
						},
						assignee: {
							href: '/api/v3/users/1',
							title: 'Bal Bahadur Pun',
						},
						project: { },
					},
				}],
			])('sets the "failed-fetching-workpackages" state on invalid responses', async (testCase) => {
				axios.get
					.mockImplementationOnce(() => Promise.resolve({
						status: 200,
						data: testCase,
					}))
					// mock for color requests, it should not fail because of the missing mock
					.mockImplementation(() => Promise.resolve({ status: 200, data: [] }))
				await wrapper.vm.setFileInfos(singleFileInfo)
				expect(wrapper.vm.state).toBe(STATE.FAILED_FETCHING_WORKPACKAGES)
			})

			it('sets the "ok" state on empty response', async () => {
				axios.get
					.mockImplementation(() => Promise.resolve({ status: 200, data: [] }))
				await wrapper.vm.setFileInfos(singleFileInfo)
				expect(wrapper.vm.state).toBe(STATE.OK)
			})
			it('sets the "error" state if the admin config is not okay', async () => {
				const returnValue = { isAdmin: false }
				getCurrentUser.mockReturnValue(returnValue)
				const wrapper = mountWrapper()
				axios.get
					.mockImplementation(() => Promise.resolve({ status: 200, data: [] }))
				await wrapper.setData({
					isAdminConfigOk: false,
				})
				await wrapper.vm.setFileInfos(singleFileInfo)
				expect(wrapper.vm.state).toBe(STATE.ERROR)
				expect(wrapper).toMatchSnapshot()
			})

			it('sets the "error" state if the admin config is not okay', async () => {
				const returnValue = { isAdmin: false }
				getCurrentUser.mockReturnValue(returnValue)
				const wrapper = mountWrapper()
				axios.get
					.mockImplementation(() => Promise.resolve({ status: 200, data: [] }))
				await wrapper.setData({
					isAdminConfigOk: false,
				})
				await wrapper.vm.setFileInfos(singleFileInfo)
				expect(wrapper.vm.state).toBe(STATE.ERROR)
				expect(wrapper).toMatchSnapshot()
			})

			it('should set workpackages to alreadylinked', async () => {
				wrapper = mountWrapper()
				axiosGetSpy = jest.spyOn(axios, 'get')
					.mockImplementationOnce(() => Promise.resolve({
						status: 200,
						data: [{
							id: 123,
							subject: 'my task',
							_links: {
								status: {
									href: '/api/v3/statuses/12',
									title: 'open',
								},
								type: {
									href: '/api/v3/types/6',
									title: 'Task',
								},
								assignee: {
									href: '/api/v3/users/1',
									title: 'Bal Bahadur Pun',
								},
								project: { title: 'a big project' },
							},
						},
						{
							id: 123,
							subject: 'my task',
							_links: {
								status: {
									href: '/api/v3/statuses/12',
									title: 'open',
								},
								type: {
									href: '/api/v3/types/6',
									title: 'Task',
								},
								assignee: {
									href: '/api/v3/users/1',
									title: 'Bal Bahadur Pun',
								},
								project: { title: 'a big project' },
							},
						}],
					}))
					.mockImplementation(() => Promise.resolve({
						status: 200,
						data: [],
					}))
				await wrapper.vm.setFileInfos(singleFileInfo)
				expect(axiosGetSpy).toBeCalledWith(
					'http://localhost/apps/integration_openproject/work-packages?fileId=123',
					{}
				)
				expect(wrapper.vm.state).toBe(STATE.OK)
				expect(wrapper.vm.alreadyLinkedWorkPackage.length).toBe(2)
			})
		})

		describe('multiple files selected', () => {
			it('sets the "error" state if the admin config is not okay', async () => {
				const returnValue = { isAdmin: false }
				getCurrentUser.mockReturnValue(returnValue)
				const wrapper = mountWrapper()
				axios.get
					.mockImplementation(() => Promise.resolve({ status: 200, data: [] }))
				await wrapper.setData({
					isAdminConfigOk: false,
				})
				await wrapper.vm.setFileInfos(multipleFileInfo)
				expect(wrapper.vm.state).toBe(STATE.ERROR)
				expect(wrapper).toMatchSnapshot()
			})

			it('should not fetch any workpackages', async () => {
				await wrapper.setData({
					isAdminConfigOk: false,
				})
				await wrapper.vm.setFileInfos(multipleFileInfo)
				await expect(axios.get).toBeCalledTimes(0)
			})
		})

		describe('onSave', () => {
			it('should closed the modal', async () => {
				await wrapper.vm.onSaved()
				expect(wrapper.find(ncModalStubSelector).exists()).toBeFalsy()
			})

			it('should empty "alreadyLinkedWorkPackage", "fileInfos" and close modal', async () => {
				await wrapper.setData({
					fileInfos: singleFileInfo,
					alreadyLinkedWorkPackage: [{
						fileId: 123,
						id: '1',
						subject: 'Organize work-packages',
						project: 'test',
						projectId: '15',
						statusTitle: 'in-progress',
						typeTitle: 'task',
						assignee: 'test',
						statusCol: 'blue',
						typeCol: 'red',
						picture: '/server/index.php/apps/integration_openproject/avatar?userId=1&userName=System',
					}],
				})
				await wrapper.vm.onSaved()
				expect(wrapper.find(ncModalStubSelector).exists()).toBeFalsy()
			})
		})
	})
})
function mountWrapper() {
	return mount(LinkMultipleFilesModal, {
		localVue,
		attachTo: document.body,
		mocks: {
			t: (app, msg) => msg,
			generateUrl() {
				return '/'
			},
		},
		stubs: {
			SearchInput: true,
			NcModal: true,
			EmptyContent: true,
		},
		data: () => ({
			show: true,
			state: STATE.OK,
			fileInfos: [],
			alreadyLinkedWorkPackage: [],
			isAdminConfigOk: true,
			searchOrigin: WORKPACKAGES_SEARCH_ORIGIN.LINK_MULTIPLE_FILES_MODAL,
		}),
	})
}
