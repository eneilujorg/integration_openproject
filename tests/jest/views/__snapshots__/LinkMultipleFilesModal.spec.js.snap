// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkMultipleFilesModal.vue fetch workpackages multiple files selected sets the "error" state if the admin config is not okay 1`] = `
<div class="multiple-link-modal-container">
  <ncmodal-stub title="" slideshowdelay="5000" enableswipe="true" size="normal" canclose="true" container="body" closebuttoncontained="true" additionaltrapelements="" inlineactions="0">
    <div class="multiple-link-modal-content">
      <div class="multiple-link-modal-inside-content">
        <h2>
          Link to work package
        </h2>
        <!---->
        <emptycontent-stub state="error" errormessage="" fileinfo="[object Object]" ismultipleworkpackagelinking="true" id="openproject-empty-content"></emptycontent-stub>
      </div>
    </div>
  </ncmodal-stub>
</div>
`;

exports[`LinkMultipleFilesModal.vue fetch workpackages single file selected sets the "error" state if the admin config is not okay 1`] = `
<div class="multiple-link-modal-container">
  <ncmodal-stub title="" slideshowdelay="5000" enableswipe="true" size="normal" canclose="true" container="body" closebuttoncontained="true" additionaltrapelements="" inlineactions="0">
    <div class="multiple-link-modal-content">
      <div class="multiple-link-modal-inside-content">
        <h2>
          Link to work package
        </h2>
        <!---->
        <emptycontent-stub state="error" errormessage="" fileinfo="[object Object]" ismultipleworkpackagelinking="true" id="openproject-empty-content"></emptycontent-stub>
      </div>
    </div>
  </ncmodal-stub>
</div>
`;

exports[`LinkMultipleFilesModal.vue fetch workpackages single file selected sets the "error" state if the admin config is not okay 2`] = `
<div class="multiple-link-modal-container">
  <ncmodal-stub title="" slideshowdelay="5000" enableswipe="true" size="normal" canclose="true" container="body" closebuttoncontained="true" additionaltrapelements="" inlineactions="0">
    <div class="multiple-link-modal-content">
      <div class="multiple-link-modal-inside-content">
        <h2>
          Link to work package
        </h2>
        <!---->
        <emptycontent-stub state="error" errormessage="" fileinfo="[object Object]" ismultipleworkpackagelinking="true" id="openproject-empty-content"></emptycontent-stub>
      </div>
    </div>
  </ncmodal-stub>
</div>
`;
